Computeserver tweak for nerd factor and readability.

som-snytt opened the pull request:
Solves the problem of two trailing closing braces, which is unaesthetic.
Keeps for( because I'm a traditionalist, but future{ so that }) is easy
on the eyes. Nested parens are OK, of course, but the thorny point on the
brace keeps its neighbor visually at a distance.

som-snytt committed:
Computeserver tweak for nerd factor and readability.

Solves the problem of two trailing closing braces, which is unaesthetic.
Keeps for( because I'm a traditionalist, but future{ so that }) is easy
on the eyes. Nested parens are OK, of course, but the thorny point on the
brace keeps its neighbor visually at a distance.

jsuereth commented:
I'm of the opinion that extra `()` belong in LISP derivatives and `;` for Java.  If you can drop both, why not!

I prefer the following because it feels more do-notation/monadic

``` scala
for {
  i <- 1 to n
  f = future { processor(i) } 
} f onComplete { _ => doneLatch.countDown() }
```

Although the extra {} for closures are ... well extra.

Maybe we should make a DSL -

``` scala
def futured[A,B](f: A => B): A => Future[B] = { in => future(f(in)) }
def countWhenDone[A](f: Future[A]) = f onComplete { _ => doneLatch.countDown() }

// Dotless notation!!!!!!!!!!!!!ONE!
1 to 10 map futured(processor) foreach countWhenDone
```

And then go home happy, knowing we have no `{}`, `;` or `.` syntax in the "core" code.

som-snytt commented:
I promise to read the book!

I'll retweak it so my fingers understand.

som-snytt committed:
Computeserver: not nearly nerdy or readable enough.

Besides the line already discussed, some additional LOC for exception
handling, possibly inspired by //FIXME what if this throws? in the library.
Also, anticipating what a learner like myself might want to do, added an
arg for "processor" count, therefore changed App to main.

With the advent of the puzzlers site and ls.implicit.ly, maybe we need
eg.implicit.ly. I always forget what e.g. stands for: exempli gratia,
which wikipedia says does not mean "free sample."

jsuereth commented on examples/src/main/scala/computeserver.scala:
```diff
@@ -3,42 +3,53 @@ package examples
         val job = openJobs.read
         printf("processor %d read a job\n", i)
-        job.ret(job.task)
+        try {
```
Just another style suggestion:

``` scala
try job ret job.task
catch {
  case x => job err x
}
```

jsuereth commented:
yeah, this is looking a lot better.  I wish I had time to help get rid of all the `get`s in there (was that a pun? probably not).

Random comment:  Sunny Munchkins!

I'll merge it shortly after I come up for air from other tasks.

som-snytt committed:
Rescue a few more dots and parens. Also, simplify counting down.

If you love something, set it free.  If the dots and parens don't come
back, they were never really meant to be there in the first place.

som-snytt commented:
I made the recommended humanitarian change, which saves not only dots and parens but eyesight.

In addition, I was concerned that I may have forgotten a latch somewhere.  Can Scala make that easier for me?

Also, I forgot how I compiled and ran this thing a couple of weeks ago.  There should be a simple way to do this out of the box for a total nuli (aka noob).  Maybe scala someExample is enough -- just make scripting scala handle arbitrary program text.  Or, maybe scala -example is super duper smart for me.

som-snytt commented:
I see the diff doesn't diff the previous two commits, but the entire branch.

The counting change is the counting method plus the implicit to wrap onComplete with the latch countdown.  The trait is named for clarity, but perhaps that could be done structurally.

I'm still at the stage where I ask, Will this syntax compile?  I want that feeling to go away, but I also know that someday I will reflect nostalgically upon these pre-Scala-3.0 days.

jsuereth commented:
Yeah, I think we can do something to help make that easier.   Give me some time to think, I'll see what I can do.

As far as testing, yeah I want to make it easier.  I'm working on something...
